types/mithril/index.d.ts

```diff
@@ -1,863 +1,268 @@
-			context: Context,
-			vdom: VirtualElement
-		): void;
+	export interface Route {
+		/** Creates application routes and mounts Components and/or RouteResolvers to a DOM element. */
+		(element: Element, defaultRoute: string, routes: RouteDefs): void;
+		/** Returns the last fully resolved routing path, without the prefix. */
+		get(): string;
+		/** Redirects to a matching route or to the default route if no matching routes can be found. */
+		set(route: string, data?: any, options?: RouteOptions): void;
+		/** Defines a router prefix which is a fragment of the URL that dictates the underlying strategy used by the router. */
+		prefix(urlFragment: string): void;
+		/** This method is meant to be used in conjunction with an <a> Vnode's oncreate hook. */
+		link(vnode: Vnode<any, any>): (e?: Event) => any;
+		/** Returns the named parameter value from the current route. */
+		param(name: string): string;
+		/** Gets all route parameters. */
+		param(): any;
 	}
 
-	/**
-	* This represents the attributes available for configuring virtual elements,
-	* beyond the applicable DOM attributes.
-	*
-	* @see m
-	*/
-	interface Attributes {
-		/**
-		* The class name(s) for this virtual element, as a space-separated list.
-		*/
-		className?: string;
+	export interface Mount {
+		/** Mounts a component to a DOM element, enabling it to autoredraw on user events. */
+		(element: Element, component: ComponentTypes<any, any>): void;
+		/** Unmounts a component from a DOM element. */
+		(element: Element, component: null): void;
+	}
 
-		/**
-		* The class name(s) for this virtual element, as a space-separated list.
-		*/
-		class?: string;
+	/** Creates an event handler which takes the value of the specified DOM element property and calls a function with it as the argument. */
+	export type WithAttr = (name: string, callback: (value: any) => any, thisArg?: any) => (e: { currentTarget: any, [p: string]: any }) => void;
 
-		/**
-		* A custom, low-level configuration in case this element needs special
-		* cleanup after removal from the tree.
-		*
-		* @see ElementConfig
-		*/
-		config?: ElementConfig;
-
-		/**
-		* A key to optionally associate with this element.
-		*/
-		key?: string | number;
+	/** Returns an object with key/value pairs parsed from a string of the form: ?a=1&b=2 */
+	export type ParseQueryString = (queryString: string) => { [p: string]: any };
 
-		/**
-		* Any other virtual element properties, including attributes and event
-		* handlers.
-		*/
-		[property: string]: any;
-	}
+	/** Turns the key/value pairs of an object into a string of the form: a=1&b=2 */
+	export type BuildQueryString = (values: { [p: string]: any }) => string;
 
-	/**
-	* The basis of a Mithril controller instance.
-	*/
-	interface Controller {
-		/**
-		* An optional handler to call when the associated virtual element is
-		* destroyed.
-		*
-		* @param evt An associated event.
-		*/
-		onunload?(evt: Event): any;
+	export interface RequestOptions<T> {
```

Comment:
I'm taking a stab at moving the type definitions into their respective files since the functions have to be declared outside of the Mithril namespace. However if I move `RequestOptions` into `request.d.ts` I can't re-export the type from the Mithril namespace without doing something like:

    import { RequestOptions as _RequestOptions} from 'mithril/request';

    declare namespace Mithril {
        //
        export type RequestOptions<T> = _RequestOptions<T>;

Which isn't great for editor hinting. Any better ideas?